File: packages/acp-py/maia_acp/registry.py
"""Agent registry and lightweight routing helpers for ACP."""

from __future__ import annotations

from dataclasses import dataclass

from maia_acp.types import ACPCapabilities, AgentPresence, MessageIntent


def _score_agent(capabilities: ACPCapabilities) -> float:
    presence = capabilities.presence
    availability = presence.availability if presence else "available"
    task_count = int(presence.active_task_count or 0) if presence else 0
    capacity = max(1, int(capabilities.max_concurrent_tasks or 1))
    load_penalty = min(0.6, task_count / capacity)
    availability_bonus = {
        "available": 1.0,
        "focused": 0.7,
        "busy": 0.35,
        "offline": 0.0,
    }.get(availability or "available", 0.5)
    return availability_bonus - load_penalty


@dataclass
class ACPAgentRegistry:
    _capabilities: dict[str, ACPCapabilities]

    def __init__(self) -> None:
        self._capabilities = {}

    def upsert_capabilities(self, capabilities: ACPCapabilities) -> None:
        previous = self._capabilities.get(capabilities.agent_id)
        if previous is None:
            self._capabilities[capabilities.agent_id] = capabilities
            return
        merged_presence = (
            previous.presence.model_copy(
                update=capabilities.presence.model_dump(exclude_none=True)
            )
            if previous.presence and capabilities.presence
            else (capabilities.presence or previous.presence)
        )
        self._capabilities[capabilities.agent_id] = capabilities.model_copy(update={"presence": merged_presence})

    def update_presence(self, agent_id: str, presence: AgentPresence) -> None:
        previous = self._capabilities.get(agent_id)
        if previous is None:
            self._capabilities[agent_id] = ACPCapabilities(
                agent_id=agent_id,
                name=agent_id.replace("agent://", ""),
                skills=[],
                presence=presence,
            )
            return
        merged = previous.presence.model_copy(update=presence.model_dump(exclude_none=True)) if previous.presence else presence
        self._capabilities[agent_id] = previous.model_copy(update={"presence": merged})

    def get_capabilities(self, agent_id: str) -> ACPCapabilities | None:
        return self._capabilities.get(agent_id)

    def list_agents(self) -> list[ACPCapabilities]:
        return list(self._capabilities.values())

    def list_available_agents(self, intent: MessageIntent | None = None) -> list[ACPCapabilities]:
        candidates = []
        for agent in self._capabilities.values():
            if agent.presence and agent.presence.availability == "offline":
                continue
            if intent and agent.accepts_intents and intent not in agent.accepts_intents:
                continue
            candidates.append(agent)
        return sorted(candidates, key=_score_agent, reverse=True)

    def resolve_recipient(
        self,
        *,
        to: str | None = None,
        intent: MessageIntent | None = None,
        exclude_agent_id: str | None = None,
    ) -> ACPCapabilities | None:
        if to and to != "agent://broadcast":
            return self._capabilities.get(to)
        for agent in self.list_available_agents(intent):
            if agent.agent_id != exclude_agent_id:
                return agent
        return None

File: packages/acp-js/src/registry.ts
import type {
  ACPCapabilities,
  ACPRegistryLike,
  AgentPresence,
  MessageIntent,
} from "./types";

function scoreAgent(capabilities: ACPCapabilities): number {
  const presence = capabilities.presence;
  const availability = presence?.availability ?? "available";
  const taskCount = Number(presence?.active_task_count ?? 0);
  const capacity = Math.max(1, Number(capabilities.max_concurrent_tasks ?? 1));
  const loadPenalty = Math.min(0.6, taskCount / capacity);
  const availabilityBonus =
    availability === "available"
      ? 1
      : availability === "focused"
        ? 0.7
        : availability === "busy"
          ? 0.35
          : 0;
  return availabilityBonus - loadPenalty;
}

class ACPAgentRegistry implements ACPRegistryLike {
  private capabilities = new Map<string, ACPCapabilities>();

  upsertCapabilities(capabilities: ACPCapabilities): void {
    const previous = this.capabilities.get(capabilities.agent_id);
    this.capabilities.set(capabilities.agent_id, {
      ...previous,
      ...capabilities,
      presence: {
        ...(previous?.presence ?? {}),
        ...(capabilities.presence ?? {}),
      },
    });
  }

  updatePresence(agentId: string, presence: AgentPresence): void {
    const previous = this.capabilities.get(agentId);
    if (!previous) {
      this.capabilities.set(agentId, {
        agent_id: agentId,
        name: agentId.replace("agent://", ""),
        skills: [],
        presence,
      });
      return;
    }
    this.capabilities.set(agentId, {
      ...previous,
      presence: {
        ...(previous.presence ?? {}),
        ...presence,
      },
    });
  }

  getCapabilities(agentId: string): ACPCapabilities | undefined {
    return this.capabilities.get(agentId);
  }

  listAgents(): ACPCapabilities[] {
    return Array.from(this.capabilities.values());
  }

  listAvailableAgents(intent?: MessageIntent): ACPCapabilities[] {
    return this.listAgents()
      .filter((agent) => {
        if (agent.presence?.availability === "offline") {
          return false;
        }
        if (!intent) {
          return true;
        }
        if (!agent.accepts_intents || agent.accepts_intents.length === 0) {
          return true;
        }
        return agent.accepts_intents.includes(intent);
      })
      .sort((left, right) => scoreAgent(right) - scoreAgent(left));
  }

  resolveRecipient(input: {
    to?: string;
    intent?: MessageIntent;
    excludeAgentId?: string;
  }): ACPCapabilities | undefined {
    const direct = String(input.to || "").trim();
    if (direct && direct !== "agent://broadcast") {
      return this.getCapabilities(direct);
    }
    return this.listAvailableAgents(input.intent).find(
      (agent) => agent.agent_id !== input.excludeAgentId,
    );
  }
}

export { ACPAgentRegistry };
